zipline/_protocol.pyx

```diff
@@ -76,6 +79,289 @@ cdef class BarData:
+                    asset: self.data_portal.get_spot_value(
+                        asset,
+                        fields,
+                        self.simulation_dt_func(),
+                        self.data_frequency)
+                    for asset in assets
+                    }, index=assets, name=fields)
+            else:
+                # both assets and fields are iterable
+                data = {}
+
+                for field in fields:
+                    series = pd.Series(data={
+                        asset: self.data_portal.get_spot_value(
+                            asset,
+                            field,
+                            self.simulation_dt_func(),
+                            self.data_frequency)
+                        for asset in assets
+                        }, index=assets, name=field)
+
+                    data[field] = series
+
+                return pd.DataFrame(data)
+
+    def can_trade(self, assets):
+        """
+        For the given asset or iterable of assets, returns true if the asset
+        is alive at the current simulation time and there is a known last
+        price.
+
+        Parameters
+        ----------
+        assets: Asset or iterable of assets
+
+        Returns
+        -------
+        boolean or Series of booleans, indexed by asset.
+        """
+        dt = self.simulation_dt_func()
+        data_portal = self.data_portal
+
+        if isinstance(assets, Asset):
+            return self._can_trade_for_asset(assets, dt, data_portal)
+        else:
+            return pd.Series(data={
+                asset: self._can_trade_for_asset(asset, dt, data_portal)
+                for asset in assets
+            })
+
+    cdef _can_trade_for_asset(self, asset, dt, data_portal):
+        if asset.start_date > dt:
+            return False
+
+        if asset.end_date < dt:
+            return False
+
+        # is there a last price?
+        return not np.isnan(
+            data_portal.get_spot_value(asset, "price", dt, self.data_frequency)
+        )
+
+    def is_stale(self, assets):
+        """
+        For the given asset or iterable of assets, returns true if the asset
+        is alive and there is no trade data for the current simulation time.
+
+        If the asset has never traded, returns False.
+
+        Parameters
+        ----------
+        assets: Asset or iterable of assets
+
+        Returns
+        -------
+        boolean or Series of booleans, indexed by asset.
+        """
+        dt = self.simulation_dt_func()
+        data_portal = self.data_portal
+
+        if isinstance(assets, Asset):
+            return self._is_stale_for_asset(assets, dt, data_portal)
+        else:
+            return pd.Series(data={
+                asset: self._is_stale_for_asset(asset, dt, data_portal)
+                for asset in assets
+            })
+
+    cdef _is_stale_for_asset(self, asset, dt, data_portal):
+        if asset.start_date > dt:
```

Comment:
Can you put a docstring here about the semantics of `stale`? `stale` means no trade for  data for the current bar, correct?

Do we need the `asset.start_date` check here? Returning `False` when there is no data seems counter intuitive, and this function is only needed if and only if the stock is 'active' correct?

That is, we are incurring extra dt comps that may be unneeded.